File: skills/JuliusBrussee--caveman/packages/middleware/python/caveman_middleware/google.py
"""Google GenAI 2.22 native HTTP transports and native AFC registration.

Install the transport through HttpOptions.httpx_client/httpx_async_client. Wrap
the existing client or chat to register the recovery executor in its native AFC
loop. The SDK keeps provider auth, Vertex configuration, retries and history.
"""
import codecs
import contextvars
import functools
import json
import re
import uuid
from urllib.parse import urlsplit

try:
    import httpx
    from google import genai
    from google.genai import types
    from google.genai.chats import Chat, AsyncChat
except ModuleNotFoundError as error:
    raise ImportError("Install caveman-middleware[google] to use the Google adapter") from error

from caveman_cloud.middleware import Adapter, Candidate, MiddlewareRuntime, AsyncMiddlewareRuntime, sha256
from ._native import Attempt, manifest, owner, plain
from ._versions import supports_framework
from ._usage import usage
from ._google_wire import parse, patch

_invocation = contextvars.ContextVar("caveman_google_invocation", default=None)


def _get(value, name, default=None):
    if plain(value):
        pieces = name.split("_")
        alias = pieces[0] + "".join(piece.title() for piece in pieces[1:])
        return value.get(name, value.get(alias, default))
    return getattr(value, name, default)


def _permits(config):
    choice = _get(_get(config, "tool_config"), "function_calling_config")
    return not (_get(config, "cached_content") or _get(config, "response_schema") or _get(config, "response_json_schema")) and _get(config, "response_mime_type") in (None, "text/plain") and _get(choice, "mode") in (None, "AUTO", types.FunctionCallingConfigMode.AUTO)


def _bind(config, runtime, scope, *, asynchronous=False):
    context = {"binding": None, "logical_call_id": str(uuid.uuid4()), "streams": set()}
    if _invocation.get() is not None:
        return config, _invocation.get()
    if type(config) not in (dict, types.GenerateContentConfig):
        return config, context
    tools = _get(config, "tools")
    afc = _get(config, "automatic_function_calling")
    maximum = _get(afc, "maximum_remote_calls")
    if runtime.mode != "compress" or owner.get() is not None or not isinstance(tools, (list, tuple)) or not any(callable(t) for t in tools) or _get(afc, "disable") or (maximum is not None and maximum <= 0) or not _permits(config):
        return config, context
    for tool in tools:
        if callable(tool) and getattr(tool, "__name__", None) == "caveman_retrieve":
            return config, context
        if not callable(tool) and _get(tool, "function_declarations"):
            # Preserve the SDK's mixed callable/declaration handling.
            return config, context
    binding = runtime.recovery(scope)

    def caveman_retrieve(handle: str, offset: int = 0, limit: int = 262144, query: str = "") -> dict:
        return binding.execute({"handle": handle, "offset": offset, "limit": limit, **({"query": query} if query else {})})

    async def async_retrieve(handle: str, offset: int = 0, limit: int = 262144, query: str = "") -> dict:
        return await binding.execute({"handle": handle, "offset": offset, "limit": limit, **({"query": query} if query else {})})

    function = async_retrieve if asynchronous else caveman_retrieve
    function.__name__, function.__doc__ = binding.name, binding.description
    # The SDK generates the tool declaration from this exact Python callable.
    # Its native schema is separately attested; runtime argument validation is
    # still authoritative for limits and handle syntax.
    declarations = [types.FunctionDeclaration.from_callable_with_api_option(callable=function, api_option=api, use_json_schema=True).model_dump(mode="json", exclude_none=True) for api in ("GEMINI_API", "VERTEX_AI")]
    context.update(binding=binding, declarations=declarations, overhead=json.dumps(declarations[0], ensure_ascii=False, separators=(",", ":")))
    changed = {"tools": [*tools, function]}
    return ({**config, **changed} if plain(config) else config.model_copy(update=changed)), context


def _scope_iterator(iterator, context):
    try:
        while True:
            token = _invocation.set(context)
            try:
                value = next(iterator)
            except StopIteration:
                return
            finally:
                _invocation.reset(token)
            yield value
    finally:
        if hasattr(iterator, "close"):
            iterator.close()
        for stream in list(context.get("streams", ())):
            if isinstance(stream, _SyncStream):
                stream.close()


async def _scope_async_iterator(iterator, context):
    try:
        while True:
            token = _invocation.set(context)
            try:
                value = await anext(iterator)
            except StopAsyncIteration:
                return
            finally:
                _invocation.reset(token)
            yield value
    finally:
        if hasattr(iterator, "aclose"):
            await iterator.aclose()
        for stream in list(context.get("streams", ())):
            if isinstance(stream, _AsyncStream):
                await stream.aclose()


def _wrap_model(module, runtime, scope, asynchronous=False):
    generate, stream = module.generate_content, module.generate_content_stream
    if asynchronous:
        @functools.wraps(generate)
        async def generate_content(*, model, contents, config=None):
            config, context = _bind(config, runtime, scope, asynchronous=True)
            token = _invocation.set(context)
            try:
                return await generate(model=model, contents=list(contents) if type(contents) is list else contents, config=config)
            finally:
                _invocation.reset(token)

        @functools.wraps(stream)
        async def generate_content_stream(*, model, contents, config=None):
            config, context = _bind(config, runtime, scope, asynchronous=True)
            token = _invocation.set(context)
            try:
                result = await stream(model=model, contents=list(contents) if type(contents) is list else contents, config=config)
            finally:
                _invocation.reset(token)
            return _scope_async_iterator(result, context)
    else:
        @functools.wraps(generate)
        def generate_content(*, model, contents, config=None):
            config, context = _bind(config, runtime, scope)
            token = _invocation.set(context)
            try:
                return generate(model=model, contents=list(contents) if type(contents) is list else contents, config=config)
            finally:
                _invocation.reset(token)

        @functools.wraps(stream)
        def generate_content_stream(*, model, contents, config=None):
            config, context = _bind(config, runtime, scope)
            return _scope_iterator(stream(model=model, contents=list(contents) if type(contents) is list else contents, config=config), context)
    module.generate_content, module.generate_content_stream = generate_content, generate_content_stream


def with_caveman_google(client, *, runtime, scope):
    """Register native model AFC on an existing, transport-configured Client.

    This is instance-scoped registration on public methods. Construct the client
    with the transports below; existing auth and client options remain native.
    Use with_caveman_google_chat for a Chat, which owns a separate AFC loop.
    """
    if not isinstance(client, genai.Client) or not isinstance(runtime, MiddlewareRuntime):
        raise TypeError("Expected a Google Client and synchronous MiddlewareRuntime")
    if not supports_framework(runtime, ("google-genai", "2.22", "3")):
        return client
    _wrap_model(client.models, runtime, scope)
    _wrap_model(client.aio.models, runtime.as_async(), scope, True)
    return client


def with_caveman_google_chat(chat, *, runtime, scope, config=None):
    """Register recovery in an existing native Chat/AsyncChat's own AFC loop.

    Pass the original default config here, or supply config on each send. No
    private chat configuration/history is read or modified.
    """
    asynchronous = isinstance(chat, AsyncChat)
    if not isinstance(chat, (Chat, AsyncChat)):
        raise TypeError("Expected a native Google Chat or AsyncChat")
    if asynchronous and isinstance(runtime, MiddlewareRuntime):
        runtime = runtime.as_async()
    if not isinstance(runtime, AsyncMiddlewareRuntime if asynchronous else MiddlewareRuntime):
        raise TypeError("Match the sync/async runtime to the chat")
    if not supports_framework(runtime, ("google-genai", "2.22", "3")):
        return chat
    send, stream, default = chat.send_message, chat.send_message_stream, config
    if asynchronous:
        @functools.wraps(send)
        async def send_message(message, config=None):
            selected, context = _bind(config if config is not None else default, runtime, scope, asynchronous=True)
            token = _invocation.set(context)
            try:
                return await send(message, config=selected)
            finally:
                _invocation.reset(token)

        @functools.wraps(stream)
        async def send_message_stream(message, config=None):
            selected, context = _bind(config if config is not None else default, runtime, scope, asynchronous=True)
            token = _invocation.set(context)
            try:
                result = await stream(message, config=selected)
            finally:
                _invocation.reset(token)
            return _scope_async_iterator(result, context)
    else:
        @functools.wraps(send)
        def send_message(message, config=None):
            selected, context = _bind(config if config is not None else default, runtime, scope)
            token = _invocation.set(context)
            try:
                return send(message, config=selected)
            finally:
                _invocation.reset(token)

        @functools.wraps(stream)
        def send_message_stream(message, config=None):
            selected, context = _bind(config if config is not None else default, runtime, scope)
            return _scope_iterator(stream(message, config=selected), context)
    chat.send_message, chat.send_message_stream = send_message, send_message_stream
    return chat


def _selected(body, strings):
    leaves, names = [], set()
    for ci, content in enumerate(body["contents"]):
        if not plain(content) or type(content.get("parts")) is not list:
            continue
        for part in content["parts"]:
            if plain(part) and plain(part.get("functionCall")) and type(part["functionCall"].get("name")) is str:
                names.add(part["functionCall"]["name"])
        for pi, part in enumerate(content["parts"]):
            if not plain(part) or part.get("thought") or part.get("thoughtSignature") or not plain(part.get("functionResponse")):
                continue
            result = part["functionResponse"]
            if type(result.get("name")) is not str or result["name"] not in names or result["name"] == "caveman_retrieve" or not plain(result.get("response")) or "error" in result["response"]:
                continue
            for key in ("output", "result"):
                path = ("contents", ci, "parts", pi, "functionResponse", "response", key)
                if type(result["response"].get(key)) is str and path in strings:
                    leaves.append((path, strings[path]))
    return leaves


def _prepare(request, runtime, scope):
    if owner.get() is not None or request.method != "POST" or not re.search(r":(?:generateContent|streamGenerateContent)$", request.url.path):
        return None
    context = _invocation.get() or {}
    attempt = Attempt(runtime, scope, context.get("logical_call_id") or str(uuid.uuid4()), str(uuid.uuid4()),
                      adapter="google-sdk", reason="opaque_payload")
    if runtime.mode == "off" or not supports_framework(runtime, ("google-genai", "2.22", "3")):
        attempt.passive = True
        attempt.reason = "disabled" if runtime.mode == "off" else "unsupported_version"
        return attempt, None
    if any(name in request.headers for name in ("content-encoding", "digest", "content-digest", "content-md5", "signature", "signature-input", "x-amz-content-sha256", "dpop")):
        return attempt, None
    try:
        if len(request.content) > 2 << 20:
            return attempt, None
        text = request.content.decode("utf-8")
    except (httpx.RequestNotRead, UnicodeError):
        return attempt, None
    parsed = parse(text)
    if not parsed:
        return attempt, None
    body, strings = parsed
    if not plain(body) or type(body.get("contents")) is not list:
        return attempt, None
    if body.get("cachedContent"):
        attempt.reason = "opaque_history_reference"
        return attempt, None
    history = manifest([{key: value for key, value in body.items() if key != "contents"}, *body["contents"]])
    if history is None:
        return attempt, None
    binding = context.get("binding")
    generation = body.get("generationConfig") or {}
    tool_config, tools = body.get("toolConfig") or {}, body.get("tools") or []
    if not plain(generation) or not plain(tool_config) or type(tools) is not list:
        return attempt, None
    choice = tool_config.get("functionCallingConfig") or {}
    if not plain(choice):
        return attempt, None
    mode = choice.get("mode")
    declarations = [d for t in tools if plain(t) and type(t.get("functionDeclarations")) is list for d in t["functionDeclarations"] if plain(d) and d.get("name") == "caveman_retrieve"]
    if not (runtime.owns_binding(binding, scope) and len(declarations) == 1 and declarations[0] in context.get("declarations", []) and mode in (None, "AUTO") and not generation.get("responseSchema") and not generation.get("responseJsonSchema") and generation.get("responseMimeType") in (None, "text/plain")):
        binding = None
    leaves = _selected(body, strings)
    options = dict(scope=scope, adapter=Adapter("google-sdk", "0.1.0", "2.22.0", "google-genai-wire-v1"), manifest=history,
        model={"provider": "google", "id": re.sub(r":(?:generateContent|streamGenerateContent)$", "", request.url.path), "protocol": "google-genai"},
        candidates=[Candidate(id=f"leaf-{i}", source_id="/".join(map(str, path)), content=leaf[2]) for i, (path, leaf) in enumerate(leaves)],
        binding=binding, recovery_overhead_text=context.get("overhead"), logical_call_id=attempt.logical_call_id, attempt_id=attempt.attempt_id)
    return attempt, (text, leaves, options)


def _apply(request, state, outcome):
    attempt, plan = state
    outgoing = request
    if attempt.passive:
        return outgoing
    if plan and outcome:
        text, leaves, options = plan
        if attempt.runtime.mode == "off" or (options["binding"] is not None and
                not attempt.runtime.owns_binding(options["binding"], attempt.scope)):
            attempt.reason = "recovery_unavailable"
            return outgoing
        attempt.optimization = outcome if not outcome.replacements else None
        attempt.reason = "invalid_plan" if outcome.replacements else outcome.reason
        mapping = {f"leaf-{i}": leaf for i, (_, leaf) in enumerate(leaves)}
        if outcome.replacements and all(r["segment_id"] in mapping for r in outcome.replacements):
            changed = patch(text, [(mapping[r["segment_id"]], r["text"]) for r in outcome.replacements])
            if changed is not None:
                content = changed.encode("utf-8")
                headers = request.headers.copy()
                if "content-length" in headers:
                    headers["content-length"] = str(len(content))
                outgoing = httpx.Request(request.method, request.url, headers=headers, content=content, extensions=request.extensions)
                attempt.optimization = outcome
    try:
        attempt.wire_sha256 = sha256(outgoing.content.decode("utf-8"))
    except (httpx.RequestNotRead, UnicodeError):
        pass
    return outgoing


class _Usage:
    def __init__(self, sse):
        self.sse, self.text, self.last, self.overflow = sse, "", None, False
        self.terminal = False
        self.decoder = codecs.getincrementaldecoder("utf-8")("replace")

    def event(self, value):
        if plain(value) and plain(value.get("usageMetadata")):
            self.last = value["usageMetadata"]
        if plain(value) and type(value.get("candidates")) is list and any(plain(c) and c.get("finishReason") for c in value["candidates"]):
            self.terminal = True

    def feed(self, chunk):
        if self.overflow:
            return
        self.text += self.decoder.decode(chunk)
        if len(self.text) > (262144 if self.sse else 2 << 20):
            self.text, self.overflow = "", True
            return
        if self.sse:
            while "\n" in self.text:
                line, self.text = self.text.split("\n", 1)
                if line.startswith("data:"):
                    try:
                        self.event(json.loads(line[5:]))
                    except (TypeError, ValueError):
                        pass

    def finish(self):
        if not self.sse and not self.overflow:
            try:
                self.event(json.loads(self.text))
            except (TypeError, ValueError):
                pass
        return None if self.overflow or self.last is None else usage(self.last, not self.sse or self.terminal)


class _Observed:
    def __init__(self, stream, attempt, sse):
        self.stream, self.attempt, self.parser, self.ended = stream, attempt, _Usage(sse), False
        self.resources = (_invocation.get() or {}).get("streams")
        if self.resources is not None:
            self.resources.add(self)

    def finish(self, event):
        if not self.ended:
            self.ended = True
            if self.resources is not None:
                self.resources.discard(self)
            self.attempt.observe(event, self.parser.finish() if event == "completed" else None)


class _SyncStream(_Observed, httpx.SyncByteStream):
    def __iter__(self):
        try:
            for chunk in self.stream:
                self.parser.feed(chunk)
                yield chunk
            self.finish("completed")
        except GeneratorExit:
            self.finish("cancelled")
            raise
        except BaseException:
            self.finish("failed")
            raise
        finally:
            self.close()

    def close(self):
        try:
            self.stream.close()
        finally:
            self.finish("cancelled")


class _AsyncStream(_Observed, httpx.AsyncByteStream):
    async def __aiter__(self):
        try:
            async for chunk in self.stream:
                self.parser.feed(chunk)
                yield chunk
            self.finish("completed")
        except BaseException:
            self.finish("cancelled")
            raise
        finally:
            await self.aclose()

    async def aclose(self):
        try:
            await self.stream.aclose()
        finally:
            self.finish("cancelled")


class CavemanGoogleTransport(httpx.BaseTransport):
    """Public HTTPX transport; original-content request hooks run before it."""
    def __init__(self, *, runtime, scope, provider_base_url="https://generativelanguage.googleapis.com", transport=None):
        if not isinstance(runtime, MiddlewareRuntime):
            raise TypeError("Use MiddlewareRuntime with the synchronous transport")
        self.runtime, self.scope, self.transport = runtime, scope, transport or httpx.HTTPTransport()
        self.base = urlsplit(provider_base_url)

    def handle_request(self, request):
        url = urlsplit(str(request.url))
        if (url.scheme, url.netloc) != (self.base.scheme, self.base.netloc) or not url.path.startswith(self.base.path.rstrip("/") + "/"):
            return self.transport.handle_request(request)
        state = _prepare(request, self.runtime, self.scope)
        if state is None:
            return self.transport.handle_request(request)
        attempt, plan = state
        outgoing = _apply(request, state, self.runtime.optimize(**plan[2]) if plan else None)
        attempt.observe("dispatch_intent")
        token = owner.set(attempt)
        try:
            response = self.transport.handle_request(outgoing)
            if response.status_code >= 400:
                attempt.observe("failed")
            else:
                response.stream = _SyncStream(response.stream, attempt, "text/event-stream" in response.headers.get("content-type", ""))
            return response
        except BaseException:
            attempt.observe("failed")
            raise
        finally:
            owner.reset(token)

    def close(self):
        self.transport.close()


class CavemanGoogleAsyncTransport(httpx.AsyncBaseTransport):
    """Async HTTPX transport with the runtime's bounded async execution path."""
    def __init__(self, *, runtime, scope, provider_base_url="https://generativelanguage.googleapis.com", transport=None):
        if isinstance(runtime, MiddlewareRuntime):
            runtime = runtime.as_async()
        if not isinstance(runtime, AsyncMiddlewareRuntime):
            raise TypeError("Use AsyncMiddlewareRuntime with the asynchronous transport")
        self.runtime, self.scope, self.transport = runtime, scope, transport or httpx.AsyncHTTPTransport()
        self.base = urlsplit(provider_base_url)

    async def handle_async_request(self, request):
        url = urlsplit(str(request.url))
        if (url.scheme, url.netloc) != (self.base.scheme, self.base.netloc) or not url.path.startswith(self.base.path.rstrip("/") + "/"):
            return await self.transport.handle_async_request(request)
        state = _prepare(request, self.runtime, self.scope)
        if state is None:
            return await self.transport.handle_async_request(request)
        attempt, plan = state
        outgoing = _apply(request, state, await self.runtime.optimize(**plan[2]) if plan else None)
        attempt.observe("dispatch_intent")
        token = owner.set(attempt)
        try:
            response = await self.transport.handle_async_request(outgoing)
            if response.status_code >= 400:
                attempt.observe("failed")
            else:
                response.stream = _AsyncStream(response.stream, attempt, "text/event-stream" in response.headers.get("content-type", ""))
            return response
        except BaseException:
            attempt.observe("failed")
            raise
        finally:
            owner.reset(token)

    async def aclose(self):
        await self.transport.aclose()

File: skills/JuliusBrussee--caveman/packages/middleware/typescript/src/google.ts
import { AsyncLocalStorage } from 'node:async_hooks';
import { GoogleGenAI, Models, Chats, type GoogleGenAIOptions, type GenerateContentParameters, type GenerateContentConfig, type CallableTool, type FunctionCall, type HttpResponse, type Part, type Tool } from '@google/genai';
import { MiddlewareRuntime, sha256, type RecoveryBinding, type Scope, type Usage } from '@caveman-ai/sdk/middleware';
import { currentOwner, manifest, observe, plain, withOwner, type Attempt } from './common.js';
import { parseWire, patchWire, pathKey, type StringLeaf } from './wire.js';
import { adapterCompatible, frameworkVersion } from './compatibility.js';

export interface GoogleOptions { runtime: MiddlewareRuntime; scope: Scope }
type NativeClient = ConstructorParameters<typeof Models>[0];
type NativeRequest = Parameters<NativeClient['request']>[0];
interface Invocation { binding: RecoveryBinding | null; overhead?: string; logicalCallId: string }
const invocations = new AsyncLocalStorage<Invocation>();

function permitsRecovery(config?: GenerateContentConfig): boolean {
  const mode = config?.toolConfig?.functionCallingConfig?.mode;
  return !config?.cachedContent && !config?.responseSchema && !config?.responseJsonSchema &&
    (!config?.responseMimeType || config.responseMimeType === 'text/plain') && (!mode || mode === 'AUTO');
}

function callable(value: Tool | CallableTool): value is CallableTool {
  return 'callTool' in value && typeof value.callTool === 'function' && 'tool' in value && typeof value.tool === 'function';
}

function schema(binding: RecoveryBinding): Tool {
  return { functionDeclarations: [{ name: binding.name, description: binding.description, parametersJsonSchema: binding.inputSchema }] };
}

/** The SDK, including its AFC scheduler, invokes this real CallableTool. */
function recoveryTool(binding: RecoveryBinding): CallableTool {
  return {
    async tool() { return schema(binding); },
    async callTool(calls: FunctionCall[]): Promise<Part[]> {
      const parts: Part[] = [];
      for (const call of calls) if (call.name === binding.name) {
        const args = call.args;
        const output = args && typeof args.handle === 'string' ? await binding.execute({ ...args, handle: args.handle }) : { error: 'invalid_recovery_arguments' };
        parts.push({ functionResponse: { name: binding.name, ...(call.id ? { id: call.id } : {}), response: { output } } });
      }
      return parts;
    },
  };
}

function measured(n: unknown): number | null { return typeof n === 'number' && Number.isSafeInteger(n) && n >= 0 ? n : null; }
function usage(value: unknown): Usage | null {
  if (!plain(value) || !plain(value.usageMetadata)) return null;
  const u = value.usageMetadata, input = measured(u.promptTokenCount), output = measured(u.candidatesTokenCount);
  return { provenance: 'client_observed_sdk', complete: input !== null && output !== null, input_tokens: input, output_tokens: output,
    cache_read_tokens: measured(u.cachedContentTokenCount), cache_write_tokens: null, reasoning_tokens: measured(u.thoughtsTokenCount) };
}

function acceptsRecovery(body: Record<string, unknown>, binding: RecoveryBinding): boolean {
  if (!Array.isArray(body.tools) || body.cachedContent) return false;
  const generation = plain(body.generationConfig) ? body.generationConfig : {};
  const choice = plain(body.toolConfig) && plain(body.toolConfig.functionCallingConfig) ? body.toolConfig.functionCallingConfig : {};
  if (generation.responseSchema || generation.responseJsonSchema || (generation.responseMimeType && generation.responseMimeType !== 'text/plain') || (choice.mode && choice.mode !== 'AUTO')) return false;
  const matches = body.tools.flatMap(t => plain(t) && Array.isArray(t.functionDeclarations) ? t.functionDeclarations : []).filter(t => plain(t) && t.name === binding.name);
  return matches.length === 1 && matches[0].description === binding.description && JSON.stringify(matches[0].parametersJsonSchema) === JSON.stringify(binding.inputSchema);
}

function selected(body: Record<string, unknown>, strings: Map<string, StringLeaf>): StringLeaf[] {
  const out: StringLeaf[] = [], names = new Set<string>();
  if (!Array.isArray(body.contents)) return out;
  body.contents.forEach((content, ci) => {
    if (!plain(content) || !Array.isArray(content.parts)) return;
    for (const part of content.parts) if (plain(part) && plain(part.functionCall) && typeof part.functionCall.name === 'string') names.add(part.functionCall.name);
    content.parts.forEach((part, pi) => {
      if (!plain(part) || part.thought || part.thoughtSignature || !plain(part.functionResponse)) return;
      const result = part.functionResponse;
      if (typeof result.name !== 'string' || !names.has(result.name) || result.name === 'caveman_retrieve' || !plain(result.response) || 'error' in result.response) return;
      // Native callable tools commonly return output (TS) or result (Python).
      // Structured records and media remain records, never flattened to text.
      for (const key of ['output', 'result']) if (typeof result.response[key] === 'string') {
        const leaf = strings.get(pathKey(['contents', ci, 'parts', pi, 'functionResponse', 'response', key]));
        if (leaf) out.push(leaf);
      }
    });
  });
  return out;
}

async function prepare(request: NativeRequest, options: GoogleOptions, defaults?: GoogleGenAIOptions['httpOptions'], passiveReason?: string): Promise<[NativeRequest, Attempt | null]> {
  if (currentOwner() || request.httpMethod !== 'POST' || !/:(?:generateContent|streamGenerateContent)(?:\?|$)/.test(request.path)) return [request, null];
  const context = invocations.getStore();
  const attempt: Attempt = { runtime: options.runtime, scope: options.scope, logicalCallId: context?.logicalCallId ?? crypto.randomUUID(), attemptId: crypto.randomUUID(), optimization: null, wireSHA256: null, adapter: 'google-sdk', reason: 'opaque_payload' };
  if (options.runtime.mode === 'off' || passiveReason) {
    attempt.passive = true;
    attempt.reason = options.runtime.mode === 'off' ? 'disabled' : passiveReason ?? 'unsupported_version';
    return [request, attempt];
  }
  request.abortSignal?.throwIfAborted();
  let outgoing = request;
  const headers = new Headers({ ...defaults?.headers, ...request.httpOptions?.headers });
  // Extra bodies are merged later by the native client. Do not optimize a view
  // whose contents/tool contract could be replaced after this public seam.
  const protectedBody = defaults?.extraBody || request.httpOptions?.extraBody || ['content-encoding', 'digest', 'content-digest', 'content-md5', 'signature', 'signature-input', 'x-amz-content-sha256', 'dpop'].some(name => headers.has(name));
  if (typeof request.body === 'string' && !protectedBody) {
    const wire = parseWire(request.body);
    if (wire && plain(wire.value) && wire.value.cachedContent) attempt.reason = 'opaque_history_reference';
    if (wire && plain(wire.value) && Array.isArray(wire.value.contents) && !wire.value.cachedContent) {
      const { contents, ...envelope } = wire.value;
      const leaves = selected(wire.value, wire.strings), history = await manifest([envelope, ...contents]);
      if (history) {
        const binding = context?.binding && options.runtime.ownsBinding(context.binding, options.scope) && acceptsRecovery(wire.value, context.binding) ? context.binding : null;
        const optimization = await options.runtime.optimize({ scope: options.scope, adapter: { id: 'google-sdk', version: '0.1.0', framework_version: frameworkVersion('@google/genai') ?? 'unknown', serialization_revision: 'google-genai-native-wire-v1' },
          model: { provider: 'google', id: request.path.replace(/:(?:generateContent|streamGenerateContent).*$/, ''), protocol: 'google-genai' }, manifest: history,
          candidates: leaves.map((leaf, i) => ({ id: `leaf-${i}`, sourceId: leaf.path.join('/'), content: leaf.value })), binding,
          ...(binding ? { recoveryOverheadText: context?.overhead ?? JSON.stringify(schema(binding)) } : {}), logicalCallId: attempt.logicalCallId, attemptId: attempt.attemptId,
          ...(request.abortSignal ? { signal: request.abortSignal } : {}) });
        if (options.runtime.mode !== 'compress' && options.runtime.mode !== 'record' ||
          (binding && (!options.runtime.ownsBinding(binding, options.scope) || !acceptsRecovery(wire.value, binding)))) {
          attempt.reason = 'recovery_unavailable';
          return [request, attempt];
        }
        attempt.optimization = optimization.replacements.length ? null : optimization;
        attempt.reason = optimization.replacements.length ? 'invalid_plan' : optimization.reason;
        const patches = optimization.replacements.map(r => ({ leaf: leaves[Number(r.segment_id.slice(5))]!, replacement: r.text }));
        if (patches.length && patches.every(p => p.leaf)) {
          const body = patchWire(request.body, patches);
          if (body !== null) {
            outgoing = { ...request, body };
            attempt.optimization = optimization;
            if (headers.has('content-length')) {
              const updatedHeaders = { ...request.httpOptions?.headers };
              for (const key of Object.keys({ ...defaults?.headers, ...request.httpOptions?.headers })) if (key.toLowerCase() === 'content-length') updatedHeaders[key] = String(new TextEncoder().encode(body).length);
              outgoing.httpOptions = { ...request.httpOptions, headers: updatedHeaders };
            }
          }
        }
      }
    }
  }
  // This is the pre-auth native serialized body, not proof of a host retry's
  // final bytes: the SDK alone owns retries and any late extra-body merge.
  if (typeof outgoing.body === 'string' && !protectedBody) attempt.wireSHA256 = await sha256(outgoing.body);
  request.abortSignal?.throwIfAborted();
  return [outgoing, attempt];
}

function observedJSON(response: HttpResponse, onValue: (value: unknown) => void, onError: () => void): HttpResponse {
  return new Proxy(response, { get(target, property) {
    if (property === 'json') return async () => { try { const value = await target.json(); onValue(value); return value; } catch (error) { onError(); throw error; } };
    const value = Reflect.get(target, property, target);
    return typeof value === 'function' ? value.bind(target) : value;
  } });
}

function delegatedClient(client: NativeClient, options: GoogleOptions, defaults?: GoogleGenAIOptions['httpOptions'], passiveReason?: string): NativeClient {
  return new Proxy(client, { get(target, property) {
    if (property === 'request') return async (request: NativeRequest) => {
      const [outgoing, attempt] = await prepare(request, options, defaults, passiveReason);
      if (!attempt) return target.request(outgoing);
      observe(attempt, 'dispatch_intent');
      try {
        const response = await withOwner(attempt, () => target.request(outgoing));
        return observedJSON(response, value => observe(attempt, 'completed', usage(value)), () => observe(attempt, 'failed'));
      } catch (error) { observe(attempt, request.abortSignal?.aborted ? 'cancelled' : 'failed'); throw error; }
    };
    if (property === 'requestStream') return async (request: NativeRequest) => {
      const [outgoing, attempt] = await prepare(request, options, defaults, passiveReason);
      if (!attempt) return target.requestStream(outgoing);
      observe(attempt, 'dispatch_intent');
      const closeController = new AbortController();
      const signal = request.abortSignal ? AbortSignal.any([request.abortSignal, closeController.signal]) : closeController.signal;
      let stream: AsyncGenerator<HttpResponse>;
      try { stream = await withOwner(attempt, () => target.requestStream({ ...outgoing, abortSignal: signal })); }
      catch (error) { observe(attempt, request.abortSignal?.aborted ? 'cancelled' : 'failed'); throw error; }
      let last: Usage | null = null, ended = false, terminal = false;
      const finish = (event: 'completed' | 'failed' | 'cancelled') => { if (!ended) { ended = true; observe(attempt, event, event === 'completed' && last ? { ...last, complete: last.complete && terminal } : null); } };
      return {
        [Symbol.asyncIterator]() { return this; },
        async next(...args: [] | [unknown]) {
          try { const next = await withOwner(attempt, () => stream.next(...args));
            if (next.done) { finish('completed'); return next; }
            return { done: false as const, value: observedJSON(next.value, value => {
              last = usage(value) ?? last;
              if (plain(value) && Array.isArray(value.candidates) && value.candidates.some(c => plain(c) && c.finishReason)) terminal = true;
            }, () => finish('failed')) };
          } catch (error) { finish(request.abortSignal?.aborted ? 'cancelled' : 'failed'); throw error; }
        },
        async return(value?: unknown) { finish('cancelled'); closeController.abort(); return stream.return(value); },
        async throw(error?: unknown) { finish('failed'); closeController.abort(); return stream.throw(error); },
      };
    };
    const value = Reflect.get(target, property, target);
    return typeof value === 'function' ? value.bind(target) : value;
  } });
}

async function invocation(params: GenerateContentParameters, options: GoogleOptions): Promise<{ params: GenerateContentParameters; context: Invocation }> {
  // Native AFC appends to contents; isolate that list from the caller's history.
  const copy: GenerateContentParameters = { ...params, ...(params.config ? { config: { ...params.config } } : {}) };
  if (Array.isArray(params.contents)) copy.contents = params.contents.slice() as typeof params.contents;
  const context: Invocation = { binding: null, logicalCallId: crypto.randomUUID() };
  const tools = params.config?.tools, maximum = params.config?.automaticFunctionCalling?.maximumRemoteCalls;
  if (options.runtime.mode !== 'compress' || currentOwner() || !tools?.some(callable) || params.config?.automaticFunctionCalling?.disable || (maximum !== undefined && (!Number.isInteger(maximum) || maximum <= 0)) || !permitsRecovery(params.config)) return { params: copy, context };
  // Use each callable's actual declaration to reject name collisions. An
  // incompatible mixed native tool list retains the SDK's own error behavior.
  const declarations = await Promise.all(tools.map(t => callable(t) ? t.tool() : t));
  if (declarations.some(t => t.functionDeclarations?.some(f => f.name === 'caveman_retrieve')) || tools.some(t => !callable(t) && t.functionDeclarations?.length)) return { params: copy, context };
  const binding = options.runtime.recovery(options.scope);
  context.binding = binding; context.overhead = JSON.stringify(schema(binding));
  copy.config = { ...copy.config, tools: [...tools, recoveryTool(binding)] };
  return { params: copy, context };
}

/**
 * Native Google client with the same GoogleGenAIOptions and native Models/Chats.
 * Uses the SDK's protected ApiClient and public module constructors; it never
 * mutates a private SDK field or replaces the native automatic tool loop.
 * Already constructed Google clients have no public clone/transport injection:
 * reuse their original options here. Other modules retain the original client.
 */
export class CavemanGoogleGenAI extends GoogleGenAI {
  declare readonly models: Models;
  declare readonly chats: Chats;
  constructor(nativeOptions: GoogleGenAIOptions, options: GoogleOptions) {
    super(nativeOptions);
    const supported = options.runtime.mode !== 'off' && adapterCompatible('google');
    if (!supported && options.runtime.mode !== 'off') options.runtime.decline('unsupported_version');
    this.models = new Models(delegatedClient(this.apiClient, options, nativeOptions.httpOptions, supported ? undefined : options.runtime.mode === 'off' ? 'disabled' : 'unsupported_version'));
    this.chats = new Chats(this.models, this.apiClient);
    if (!supported) return;
    const generate = this.models.generateContent.bind(this.models), stream = this.models.generateContentStream.bind(this.models);
    this.models.generateContent = async params => { const call = await invocation(params, options); return invocations.run(call.context, () => generate(call.params)); };
    this.models.generateContentStream = async params => {
      const call = await invocation(params, options), result = await invocations.run(call.context, () => stream(call.params));
      const wrapped: Awaited<ReturnType<Models['generateContentStream']>> = { [Symbol.asyncIterator]() { return this; },
        next: (...args) => invocations.run(call.context, () => result.next(...args)),
        return: value => invocations.run(call.context, () => result.return(value)),
        throw: error => invocations.run(call.context, () => result.throw(error)),
      };
      return wrapped;
    };
  }
}
